### PR TITLE
Regional Delaunay 

### LINKED ISSUE #1049: Implement Regional Delaunay
### Proposed new feature or change:

Create Regional Delaunay triangles by projecting the grid to a plane, generating Delaunay triangles, and reverting back to the grid.

### PR DESCRIPTION
**Pull request by aaronzedwick**
Closes #1049 #1050 #1051 

## Overview

Adds `Grid.to_planar` which takes a grid and projects it to a plane, using `_point_to_plane` which converts a point on the sphere to the plane. Also adds `_point_to_sphere` which converts a point on a plane back to the sphere.
## Expected Usage

```Python
import uxarray as ux

grid_points = [[10, -10, 10, -10], [-10, -10, 10, 10]]

grid_r = ux.Grid.from_points(grid_points, method="regional_delaunay")

```

## PR Checklist

**General**
- [x] An issue is linked created and linked
- [x] Add appropriate labels
- [x] Filled out Overview and Expected Usage (if applicable) sections

**Testing**
- [x] Adequate tests are created if there is new functionality
- [x] Tests cover all possible logical paths in your function
- [x] Tests are not too basic (such as simply calling a function and nothing else)

**Documentation**
- [x] Docstrings have been added to all new functions
- [x] Docstrings have updated with any function changes
- [x] Internal functions have a preceding underscore (`_`) and have been added to `docs/internal_api/index.rst`
- [x] User functions have been added to `docs/user_api/index.rst`

**Examples**
- [x] Any new notebook examples added to `docs/examples/` folder
- [x] Clear the output of all cells before committing
- [x] New notebook files added to `docs/examples.rst` toctree
- [x] New notebook files added to new entry in `docs/gallery.yml` with appropriate thumbnail photo in `docs/_static/thumbnails/`

### CODE REVIEW

**Comment by aaronzedwick** [comment]
@philipc2 got any suggestions on how I might store this information? I am not completely certain about a solid way to do so.

**Comment by philipc2** [comment]
> @philipc2 got any suggestions on how I might store this information? I am not completely certain about a solid way to do so.

Do these variables need to be cached, or can they be used once and then scrapped? Considering the computation of these variables will be pretty quick thanks to Numba, I lean towards not storing them directly as part of a `Grid`.

**Comment by aaronzedwick** [comment]
> > @philipc2 got any suggestions on how I might store this information? I am not completely certain about a solid way to do so.
> 
> Do these variables need to be cached, or can they be used once and then scrapped? Considering the computation of these variables will be pretty quick thanks to Numba, I lean towards not storing them directly as part of a `Grid`.

Sorry, I think I miscommunication. I more meant, how should I return this information? Is using a dataset the best way, and constructing them all at once? Or should I have one function for each (node, face, edge) and return a np array?

**Comment by philipc2** [comment]
> > > @philipc2 got any suggestions on how I might store this information? I am not completely certain about a solid way to do so.
> > 
> > 
> > Do these variables need to be cached, or can they be used once and then scrapped? Considering the computation of these variables will be pretty quick thanks to Numba, I lean towards not storing them directly as part of a `Grid`.
> 
> Sorry, I think I miscommunication. I more meant, how should I return this information? Is using a dataset the best way, and constructing them all at once? Or should I have one function for each (node, face, edge) and return a np array?

I see! I'd lean towards not having a `Grid.to_planar()` method and instead having the `_point_to_plane()` and `_point_to_sphere()` called as needed, returning the variables instead of storing them as a dataset.

File: uxarray/grid/geometry.py [hunk 1]
```diff
@@ -1336,3 +1337,34 @@ def _construct_boundary_edge_indices(edge_face_connectivity):
     # If an edge only has one face saddling it than the mesh has holes in it
     edge_with_holes = np.where(edge_face_connectivity[:, 1] == INT_FILL_VALUE)[0]
     return edge_with_holes
+
+
+# @njit(cache=True)
```
Comment (philipc2) [inline uxarray/grid/geometry.py:1342]: Did you comment this out when debugging?

**Review by philipc2 — CHANGES_REQUESTED** [review]
Can you test the behavior if the points are along the antemeridian? If you want to use data from the usage example, you could try the following. 

`uxgrid_regional.subset.bounding_circle((-180, 0), 10)`

**Review by philipc2 — COMMENTED** [review]
Can you also expand upon the markdown in the notebook? I think the most important addition would be the "How It Works" section that describes the underlying algorithm. 

image

File: uxarray/grid/geometry.py (hunk 1 above)
Comment (aaronzedwick) [reply uxarray/grid/geometry.py:1342]: Oh, yeah I did.

**Comment by aaronzedwick** [comment]
@philipc2 This projection technique (sterographic) can't handle south pole points, or rather any points along the z=1 axis. I was thinking maybe using a gnomonic projection would work better and be more accurate (as this one seems to not work very well for point in polygon either, the projection is distorted and gives incorrect results when the point is close to the edge). The gnomonic projection might have the same issue at the poles though, not certain. Thoughts? Do you think gnomonic projection would give more accurate results?

**Review by rajeeja — CHANGES_REQUESTED** [review]
can you do something like `norm = np.linalg.norm(points, axis=1)` to calculate all norm at once?

also, is might faster to use set operation for filtering.

File: uxarray/grid/grid.py [hunk 2]
```diff
@@ -387,6 +390,8 @@ def from_points(
             ds = _spherical_voronoi_from_points(_points, **kwargs)
         elif method == "spherical_delaunay":
             ds = _spherical_delaunay_from_points(_points, boundary_points)
+        elif method == "regional_delaunay":
+            ds = _regional_delaunay_from_points(_points, boundary_points)
```
Comment (philipc2) [inline uxarray/grid/grid.py:394]: Please update the method parameter in the docstring. 

```
        method : str, optional
            The method to generate connectivity information. Supported methods are:

            - `'spherical_voronoi'`: Constructs a spherical Voronoi diagram.
            - `'spherical_delaunay'`: Constructs a spherical Delaunay triangulation.

            Default is `'spherical_delaunay'`.
```

**Review by philipc2 — CHANGES_REQUESTED** [review]
Notebook changes are looking great!

One small comment below.

**Review by philipc2 — APPROVED** [review]
Great work @aaronzedwick

**Review by rajeeja — APPROVED** [review]

**Pull request merged**